pyg_lib/csrc/utils/hetero_dispatch.h

```diff
@@ -0,0 +1,191 @@
+  bool filter_by_edge(const edge_t& edge) {
+    return val_.contains(get_src(edge));
+  }
+
+ private:
+  T val_;
+};
+
+// In NodeDstMode we check if destination node is in the dict
+template <typename T, typename V>
+class HeteroDispatchArg<T, V, NodeDstMode> {
+ public:
+  HeteroDispatchArg(const T& val) : val_(val) {
+    static_assert(is_c10_dict<T>::value, "Should be a c10::dict");
+  }
+
+  template <typename K>
+  V value_by_edge(const K& key) {
+    return val_.at(get_dst(key));
+  }
+
+  bool filter_by_edge(const edge_t& edge) {
+    return val_.contains(get_dst(edge));
+  }
+
+ private:
+  T val_;
+};
+
+// In EdgeMode we check if edge is in the dict
+template <typename T, typename V>
+class HeteroDispatchArg<T, V, EdgeMode> {
+ public:
+  HeteroDispatchArg(const T& val) : val_(val) {
+    static_assert(is_c10_dict<T>::value, "Should be a c10::dict");
+  }
+
+  template <typename K>
+  V value_by_edge(const K& key) {
+    return val_.at(key);
+  }
+
+  bool filter_by_edge(const edge_t& edge) { return val_.contains(edge); }
+
+ private:
+  T val_;
+};
+
+// The following will help static type checks:
+template <typename... T>
+struct is_hetero_arg : std::false_type {};
+
+// Just check inheritance, a workaround without introducing concepts
+template <typename T, typename V, typename Mode>
+struct is_hetero_arg<HeteroDispatchArg<T, V, Mode>> : std::true_type {
+  static_assert(std::is_base_of<HeteroDispatchMode, Mode>::value,
+                "Must pass a mode for dispatching");
+};
+
+// Specialize
+template <typename... Args>
+bool filter_args_by_edge(const edge_t& edge, Args&&... args) {}
+
+// Stop condition of argument filtering
+template <>
+bool filter_args_by_edge(const edge_t& edge) {
+  return true;
+}
+
+// We filter each argument individually by the given edge using a variadic
+// template
+template <typename T, typename... Args>
+bool filter_args_by_edge(const edge_t& edge, T&& t, Args&&... args) {
+  static_assert(
+      is_hetero_arg<std::remove_const_t<std::remove_reference_t<T>>>::value,
+      "args should be HeteroDispatchArg");
+  return t.filter_by_edge(edge) && filter_args_by_edge(edge, args...);
+}
+
+// Check if a callable is wrapped by std::function
+template <typename... T>
+struct is_std_function : std::false_type {};
+
+template <typename T, typename... Args>
+struct is_std_function<std::function<T(Args...)>> : std::true_type {};
+
+template <typename T>
+class HeteroDispatchOp {
+ public:
+  using result_type = typename T::result_type;
+  HeteroDispatchOp(const edge_tensor_dict_t& rowptr,
+                   const edge_tensor_dict_t& col,
+                   T op)
+      : rowptr_(rowptr), col_(col), op_(op) {
+    // Check early
+    static_assert(is_std_function<T>::value, "Must pass a function");
+  }
+
+  template <typename... Args>
+  c10::Dict<edge_t, result_type> operator()(Args&&... args) {
+    c10::Dict<edge_t, result_type> dict;
+    for (const auto& kv : rowptr_) {
+      auto edge = kv.key();
```

Comment:
I think after you rename the type, the code will be either to read. I.e `edge_t` become `EdgeType`, `result_type` become `ResultType` since they are types.
Here 
```
auto edge_type = kv.key();
```
So that we know it is a type instead of a tensor.